lib/nats.js

```diff
@@ -824,10 +843,13 @@ Client.prototype.publish = function(subject, msg, opt_reply, opt_callback) {
   } else {
     psub = 'PUB ' + subject + SPC + opt_reply + SPC;
   }
+
   this.sendCommand(psub + Buffer.byteLength(msg) + CR_LF + msg + CR_LF);
 
   if (opt_callback !== undefined) {
     this.flush(opt_callback);
+  } else if (this.closed) {
+    this.emit('error', PUB_CONN_CLOSED_ERR);
```

Comment:
This is actually much of the problem with your main read loop.  It is too large a chunk of looping code to run efficiently in a Node environment (not that the loop itself is inefficient).  Node, being single-threaded, needs the separation in order to properly handle multiple code executions.  This is why you are having to put in that cool-off timer;  you shouldn't have to do that.